### ISSUE TITLE
microlens-platform isn't sufficient, microlens also required ?

### ISSUE THREAD

**Issue by simonmichael** [aelve/microlens#73]
To get fully-featured microlens, I got the impression I only had to depend on and import microlens-platform. But the latter only exports http://hackage.haskell.org/package/microlens-platform-0.3.1.0/docs/Lens-Micro-Platform.html (and seemingly does not re-export Lens.Micro ?). For convenience I think microlens-platform should export everything including Lens.Micro (or otherwise, docs should demonstrate how users should depend on and include both microlens and microlens-platform).

**Comment by neongreen** [comment]
It reexports `Lens.Micro.GHC`, which reexports `Lens.Micro`. I should definitely clarify that in the docs :)

**Comment by simonmichael** [comment]
Ah, I see. Is it reasonable to export everything as Lens.Micro instead of Lens.Micro.Platform ? That would make it easier to transition code to microlens-platform when you discover it, or to piecemeal microlens-* when you want to trim down.

**Comment by simonmichael** [comment]
Related, is it too late to consider making "microlens" be the full-featured ready-to-go lens replacement, and "microlens-core" or something be the minimal version ? That would have got me started quicker.

**Comment by neongreen** [comment]
(Meanwhile, I've released a new version of microlens-platform that reexports `Lens.Micro` explicitly.)

**Comment by neongreen** [comment]
> Related, is it too late to consider making "microlens" be the full-featured ready-to-use lens replacement

It's probably too late, yeah, especially since

* I kinda don't want to give special status to `vector`/`text`/`unordered-containers`
* A full-featured replacement would be something like microlens-platform + microlens-contra + yet-unwritten microlens-pro with prisms and isos (and then what would be the point?)

**Comment by simonmichael** [comment]
Thanks!

**Comment by simonmichael** [comment]
The only visible change at http://hackage.haskell.org/package/microlens-platform-0.3.1.1 is that orphan instances now appear in the haddock, by the way. Is "import Lens.Micro.Platform" still going to be required when using microlens-platform ?

**Comment by neongreen** [comment]
Apparently Haddock hid the reexport. I'll try to convince it to show it.

**Comment by neongreen** [comment]
I didn't rename any modules, so, yes, you need to import Lens.Micro.Platform. You *don't* have to import Lens.Micro or depend on microlens (and you never did!).

**Comment by neongreen** [comment]
* I'll also add a line about microlens-platform to the short description of microlens (so that it would be maximally noticeable)
* Apparently hiding reexports is a feature of new Haddock; I'm installing GHC 8 right now to check whether my way of circumventing it will work

**Comment by neongreen** [comment]
Okay, I think I figured it out. The version of Haddock shipped with GHC 8 breaks on `microlens`, and Haddock silently ignores modules coming from packages that don't have docs built for them, so it skips the `Lens.Micro` export.

I tried building the docs manually and uploading them, but Hackage didn't let me.

> I'll also add a line about microlens-platform to the short description of microlens

This is done.

**Comment by neongreen** [comment]
Have you got any other ideas about making microlens-platform more visible? Anything could help!

**Comment by simonmichael** [comment]
Well, my best idea is to call it microlens. Next best, anything you can do to highlight it and point out that it's the recommended starting point for typical users of microlens, in the package descriptions of all other microlens packages, will help folks to not repeat my mistake, ie: start out with microlens, go through a few cycles of head-scratching over missing lens-like things (beginning with the very first thing one needs, makeLenses), finding and adding the appropriate microlens-X package, before realising that microlens-platform is the one true package, and import Lens.Micro.Platform the appropriate import. (And I know you've documented all this, and I probably read all about it last time, but it wasn't enough. :-).

**Comment by neongreen** [comment]
> Well, my best idea is to call it microlens.

I really *do* understand where you're coming from. “Start out with microlens, go through a few cycles of head-scratching” is horrible UX and should be prevented (almost) at any cost. (Moreover: if I documented it but you didn't find it, it's obviously my fault, not yours.)

However, the thing is that microlens was primarily intended for library writers, not for app writers (“builds in 4 seconds” is still my goal; actually, the latest version builds in 3!). In addition to other things, I'm afraid that if I rename microlens-platform to microlens and microlens to microlens-core, I might end up hurting my chances in that market. I have to say that the UX horror is a very compelling argument, and I actually thought for a while that maybe I should just do the renaming – but ultimately right now I still feel that it wouldn't be a good decision. At any rate, I'll keep this issue open in case anybody else has an opinion on the matter.

I've added a mention of microlens-platform to the README, the beginning of the description in microlens.cabal, and there's now a description of all packages in the beginning of `Lens.Micro`.
